contracts/fee/FeeHandler.sol

```diff
@@ -3,46 +3,176 @@
+        uint256 maxFeeTokenAmount = (minFeeTokenAmount * (maxPriceImpactFactor + 10000)) / 10000;
+
+        uint256 availableFeeAmount = _getAvailableFeeAmount(feeToken, buybackToken);
+
+        if (availableFeeAmount >= maxFeeTokenAmount) {
+            _buybackFees(feeToken, buybackToken, batchSize, maxFeeTokenAmount, availableFeeAmount);
+        } else if (availableFeeAmount >= minFeeTokenAmount) {
+            _buybackFees(feeToken, buybackToken, batchSize, availableFeeAmount, availableFeeAmount);
+        } else {
+            revert Errors.InsufficientFeeAmount(feeToken, buybackToken, availableFeeAmount);
+        }
+    }
+
+    function _buybackFees(
+        address feeToken,
+        address buybackToken,
+        uint256 batchSize,
+        uint256 buybackAmount,
+        uint256 availableFeeAmount
+    ) private {
+        IERC20(feeToken).safeTransfer(msg.sender, buybackAmount);
+        IERC20(buybackToken).safeTransferFrom(msg.sender, address(this), batchSize);
+        availableFeeAmount = availableFeeAmount - buybackAmount;
+        _setAvailableFeeAmount(feeToken, buybackToken, availableFeeAmount);
+    }
+
+    function _setAvailableFeeAmounts(uint256 version, address feeToken, uint256 feeAmount) private {
+        address wnt = dataStore.getAddress(Keys.WNT);
+
+        uint256 buybackGmxFactor = _getBuybackGmxFactor(version);
+        uint256 feeAmountGmx = (feeAmount * buybackGmxFactor) / 1e30;
+        uint256 feeAmountWnt = feeAmount - feeAmountGmx;
 
-        for (uint256 i; i < markets.length; i++) {
-            FeeUtils.claimFees(
-                dataStore,
-                eventEmitter,
-                markets[i],
-                tokens[i],
-                receiver
-            );
+        uint256 availableFeeAmountGmx = _getAvailableFeeAmount(feeToken, gmx) + feeAmountGmx;
+        uint256 availableFeeAmountWnt = _getAvailableFeeAmount(feeToken, wnt) + feeAmountWnt;
+
+        _setAvailableFeeAmount(feeToken, gmx, availableFeeAmountGmx);
+        _setAvailableFeeAmount(feeToken, wnt, availableFeeAmountWnt);
+    }
+
+    function _setAvailableFeeAmount(address feeToken, address buybackToken, uint256 availableFeeAmount) private {
+        dataStore.setUint(Keys.buybackAvailableFeeAmountKey(feeToken, buybackToken), availableFeeAmount);
+    }
+
+    function _getAvailableFeeAmount(address feeToken, address buybackToken) private view returns (uint256) {
+        uint256 availableFeeAmount = dataStore.getUint(Keys.buybackAvailableFeeAmountKey(feeToken, buybackToken));
+        return availableFeeAmount;
+    }
+
+    function _getBuybackGmxFactor(uint256 version) private view returns (uint256) {
+        uint256 buybackGmxFactor = dataStore.getUint(Keys.buybackGmxFactorKey(version));
+        return buybackGmxFactor;
+    }
+
+    function _getMinFeeTokenAmount(
+        address feeToken,
+        address buybackToken,
+        uint256 batchSize
+    ) private view returns (uint256) {
+        (
```

Comment:
should use https://github.com/gmx-io/gmx-synthetics/blob/main/contracts/oracle/ChainlinkPriceFeedUtils.sol#L17 instead

the returned price would already be adjusted based on the token decimals

Reply:
Thanks, I've updated to use this instead.